### ISSUE TITLE
Can't use hash-only URLs

### ISSUE THREAD

**Issue by cognivator** [47ng/nuqs#450]
## Context

What's your version of `nuqs`?

```
"nuqs": "^1.14.0"
```

Next.js information (obtained by running `next info`):

```
Operating System:
  Platform: darwin
  Arch: x64
  Version: Darwin Kernel Version 21.6.0: Sat Jun 18 17:07:25 PDT 2022; root:xnu-8020.140.41~1/RELEASE_X86_64
Binaries:
  Node: 20.5.0
  npm: 9.8.0
  Yarn: N/A
  pnpm: N/A
Relevant Packages:
  next: 13.5.6
  eslint-config-next: 13.5.6
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.2.2
Next.js Config:
  output: N/A
```

Are you using:

- ✅ The app router
- ❌ The pages router
- ❌ The `basePath` option in your Next.js config
- ❌ The experimental `windowHistorySupport` flag in your Next.js config

## Description

If the app url includes a hash parameter (e.g. "myapp.com/#anchor"), then nuqs query string modifications are faulty in a couple of ways.

## Reproduction

[codesandbox](https://codesandbox.io/p/devbox/nuqs-poc-nk7vkc?file=%2Fapp%2Fpage.tsx%3A12%2C5&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clr5laqmb00073b6h1i7uaj0i%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clr5laqmb00023b6hvgrrs1ym%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clr5laqmb00043b6hax7m27h4%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clr5laqmb00063b6htmnoht0y%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B43.5423474438259%252C56.4576525561741%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clr5laqmb00023b6hvgrrs1ym%2522%253A%257B%2522id%2522%253A%2522clr5laqmb00023b6hvgrrs1ym%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522clr5laqmb00063b6htmnoht0y%2522%253A%257B%2522id%2522%253A%2522clr5laqmb00063b6htmnoht0y%2522%252C%2522activeTabId%2522%253A%2522clr5p6cmd01q83b6h4jwj2guy%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522clr5p6cmd01q83b6h4jwj2guy%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2523hash%2522%257D%255D%257D%252C%2522clr5laqmb00043b6hax7m27h4%2522%253A%257B%2522id%2522%253A%2522clr5laqmb00043b6hax7m27h4%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522id%2522%253A%2522clr5o0v3200bo3b6hj112g2zr%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clr5o0v3200bo3b6hj112g2zr%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

Example: Steps to reproduce the behavior:

For illustration, a simple boolean parser is used, and the setter is called from a toggle that switches between true and false, starting with false. The URL sequences all start with the initial load, then show each toggle, true-false-true-false...

Adjust the initial URL, parser, and setter as indicated, then hit the toggle button several times.

URL: hash only (/#hash)
Parser: no default
Setter: set `null` when false
Url sequence should be:
- /#hash
- /#hash?p=true#hash
- /#hash#hash?p=true#hash
- /#hash#hash?p=true#hash#hash?p=true#hash
- /#hash#hash#hash#hash?p=true#hash#hash?p=true#hash

URL: query and hash (/?p=true#hash)
Parser: no default
Setter: set `null` when false
Url sequence should be:
- /?p=true#hash
- /#hash
- (as above using null when false)

URL: hash only (/#hash)
Parser: no default
Setter: set `false` when false
Url sequence should be:
- /#hash
- /#hash?p=true#hash
- /#hash?p=false#hash?p=true#hash
- /#hash?p=true#hash?p=false#hash?p=true#hash
- /#hash?p=false#hash?p=true#hash?p=false#hash?p=true#hash

URL: query and hash (/?p=true#hash)
Parser: no default
Setter: set `false` when false
Url sequence should be:
- /?p=true#hash
- /?p=false#hash
- /?p=true#hash
- (repeats properly, indefinitely)

**Comment by franky47** [comment]
Thanks a lot for this great deep-dive, looks like the error is here:

https://github.com/47ng/nuqs/blob/e158622f3611e76468a8946541282ded0a7c6da2/packages/nuqs/src/update-queue.ts#L176

I'll use your examples to get a fix out asap.

**Comment by cognivator** [comment]
That is amazingly fast response! And quick diagnosis.
I look forward to using nuqs... tried handling URL state myself, and as I'm sure you already know the details quickly multiply. Happy that nuqs is already dealing with them.! 🙂

**Comment by cognivator** [comment]
@franky47 Is there anything I should do to follow up with 1.14.1-beta.1 to help get the fix into the main release?

**Comment by franky47** [comment]
@cognivator sorry I was just making sure the release process was still working, using the beta channel as a test bench, sorry for the spam! (but at least I found a bug 😅 ).

Should be all fixed in 1.14.1.

**Comment by cognivator** [comment]
@franky47 Many thanks! I'll give it a whirl.
